### PR TITLE
fix(monitoring): route plot callbacks through Lightning loggers

### LINKED ISSUE #614: fix(monitoring): guard direct wandb.log()/wandb.Image() calls in default callbacks
## Summary

`src/utils/callbacks.py` makes direct `wandb.log()` / `wandb.Image()` calls in three callbacks that are enabled by `configs/callbacks/default.yaml`:

| Callback       | File                                          | Call sites                    |
| -------------- | --------------------------------------------- | ----------------------------- |
| `plot_loss`    | `src/utils/callbacks.py:76-77`                | `wandb.Image`, `wandb.log`    |
| `plot_pos_enc` | `src/utils/callbacks.py:134-135`              | `wandb.Image`, `wandb.log`    |
| `plot_proj`    | `src/utils/callbacks.py:257-259`              | `wandb.Image` × 2, `wandb.log` |

`configs/callbacks/default.yaml` includes `plot_pos_enc`, `plot_proj`, `plot_projii` by default.

## Why this matters now

With #598 landed, the default logger composition is `csv + tensorboard` — no W&B. But the default callbacks still hit `wandb.log()` / `wandb.Image()` directly, bypassing the Lightning logger system entirely. This means a fresh-checkout run with defaults can still trigger a W&B auth prompt or create anonymous runs, undoing part of the "W&B is opt-in" story.

Surfaced by Copilot review on #612: https://github.com/tinaudio/synth-setter/pull/612#discussion_r3108400050

## Proposed fix

Guard each direct `wandb.log` / `wandb.Image` call on `wandb.run is not None` (or equivalent check that a Wandb run is active). If no run, no-op (optionally log to the active Lightning loggers instead — but that's a bigger refactor).

Minimal version:

```python
if wandb.run is not None:
    wandb.log({"plot": wandb.Image(fig)}, step=trainer.global_step)
```

Ideal version: route through the Lightning logger pattern (`trainer.logger.experiment.log_image(...)`) so the plot shows up on whichever logger(s) the user has selected (TensorBoard, W&B, or both).

## Out of scope

- Callback composition / separate experiment configs — tracked elsewhere under #563.

Part of #563

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

Replaces direct `wandb.log()` / `wandb.Image()` calls in the three default plot callbacks with a small Lightning-logger dispatch helper. Plots now land on whichever logger(s) the user has selected — `WandbLogger.log_image` when W&B is active, `TensorBoardLogger.experiment.add_figure` when TB is active, both if both, silently skipped otherwise.

This is the "ideal" fix from #614 (not the minimal `if wandb.run is not None` guard): after #612 made W&B opt-in and the defaults became CSV + TensorBoard, the direct `wandb.log()` calls were still bypassing the Lightning logger system and could trigger W&B auth prompts on fresh checkouts. Routing through `trainer.loggers` closes that gap and also makes plots visible in TensorBoard for the default composition.

## Changes

- `src/utils/callbacks.py`
  - New `_log_figure(trainer, key, fig)` helper that iterates `trainer.loggers` and dispatches per-logger type. Loggers without image support (CSV, etc.) are silently skipped.
  - `PlotLossPerTimestep._log_plot`, `PlotPositionalEncodingSimilarity._log_plot`, and `PlotLearntProjection._log_plots` now route through the helper instead of `wandb.log`/`wandb.Image` directly.
  - Module-level `import wandb` removed — no longer needed.
- `tests/test_callbacks.py` (new) — four behavior tests using `MagicMock(spec=...)` fakes: WandbLogger-only, TensorBoardLogger-only, both-present, and the CSV-only no-op regression guard (the fresh-checkout default must stay silent).

## Test plan

- [x] `make test` passes locally (162 passed, 2 skipped).
- [x] `make format` passes (ruff, ruff format, pyright, interrogate, etc.).
- [ ] CI green on this PR.
- [ ] Manual smoke: `python src/train.py experiment=<exp> +trainer.fast_dev_run=true` with defaults → no W&B auth prompt, plots visible in TB.

## Out of scope

- Callback composition / per-experiment callback config — tracked under #563.

Closes #614
Part of #563

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

### Explicit promises

- [x] **`_log_figure` dispatches to `WandbLogger.log_image` when a WandbLogger is in `trainer.loggers`** *(Level 2 — invoked pytest against the real dispatch code with `MagicMock(spec=WandbLogger)` so `isinstance` hits the real class)*
  ```
  $ pytest tests/test_callbacks.py -v
  tests/test_callbacks.py::test_log_figure_routes_to_wandb_logger_when_only_wandb_logger_present PASSED [ 25%]
  tests/test_callbacks.py::test_log_figure_routes_to_tensorboard_logger_when_only_tensorboard_logger_present PASSED [ 50%]
  tests/test_callbacks.py::test_log_figure_dispatches_to_both_when_both_loggers_present PASSED [ 75%]
  tests/test_callbacks.py::test_log_figure_is_noop_when_no_image_capable_loggers_present PASSED [100%]
  4 passed in 1.35s
  ```

- [x] **All three plot callbacks (`PlotLossPerTimestep`, `PlotPositionalEncodingSimilarity`, `PlotLearntProjection`) delegate through `_log_figure`** *(Level 2 — instantiated each callback and invoked its `_log_plot` / `_log_plots` method with spec'd loggers; asserted the right number of `log_image` / `add_figure` calls with the right keys)*
  ```
  $ python3 -c "
  from unittest.mock import MagicMock
  from lightning.pytorch.loggers import WandbLogger, TensorBoardLogger
  from matplotlib.figure import Figure
  from src.utils.callbacks import (
      PlotLossPerTimestep, PlotPositionalEncodingSimilarity, PlotLearntProjection,
  )
  wandb_logger = MagicMock(spec=WandbLogger)
  tb_logger = MagicMock(spec=TensorBoardLogger)
  trainer = MagicMock(loggers=[wandb_logger, tb_logger], global_step=99)
  fig = MagicMock(spec=Figure); fig2 = MagicMock(spec=Figure)
  PlotLossPerTimestep()._log_plot(fig, trainer)
  PlotPositionalEncodingSimilarity()._log_plot(fig, trainer)
  PlotLearntProjection()._log_plots(fig, fig2, trainer)
  print('wandb calls:', wandb_logger.log_image.call_count)
  print('tb calls:',    tb_logger.experiment.add_figure.call_count)
  keys = [c.kwargs['key'] for c in wandb_logger.log_image.call_args_list]
  print('wandb keys:', keys)
  "
  wandb calls: 4
  tb calls:    4
  wandb keys: ['plot', 'pos_enc_similarity', 'assignment', 'value']
  ```
  4 wandb + 4 tb calls for 1 + 1 + 2 plots. Keys match: \`plot\`, \`pos_enc_similarity\`, \`assignment\`, \`value\`.

- [x] **CSV-only logger setup is silent (the fresh-checkout default after #612)** *(Level 2 — covered by `test_log_figure_is_noop_when_no_image_capable_loggers_present` in the pytest output above; passes with zero \`method_calls\` on a CSVLogger spec)*

- [x] **`src.utils.callbacks` no longer references `wandb` at import time** *(Level 3 — imported the module in a fresh interpreter and asserted \`wandb\` is not in its namespace. The previous module-level \`import wandb\` was the fresh-checkout auth-prompt risk surfaced in #614.)*
  ```
  $ python3 -c "import src.utils.callbacks as cb; assert not hasattr(cb, 'wandb'); print('OK')"
  OK
  ```

### Platform / CI

- [x] **PR metadata gate passes** — \`check-pr-metadata\` green on CI (issue #614 has type \`Bug\`, label \`monitoring\`, milestone \`monitoring v1.0.0\`, and traces to Epic #264 via Phase #563).
- [ ] **Full CI suite green** *(Level 1 — 9/11 checks passed; \`run_tests_macos (macos-latest, 3.10)\` and \`run_tests_ubuntu (ubuntu-latest, 3.11)\` still pending at verification time. Re-run \`gh pr checks 646\` once they complete.)*

### Adversarial / implicit

- [ ] **Fresh-checkout training run with defaults produces no W&B auth prompt and writes plots to the TB log dir** *(Level 1 — the original motivation of #614. SKIP: requires an experiment config + data on the runner; not attempted in this verification environment.)*
  To verify manually on a machine with training data:
  ```bash
  # No WANDB_* env vars set; defaults = CSV + TensorBoard
  unset WANDB_API_KEY WANDB_ENTITY
  python src/train.py experiment=<your-exp> +trainer.fast_dev_run=true
  # Expect: no \"wandb login\" prompt, no anonymous-run URL printed,
  # and plots appear under logs/.../tensorboard/ (viewable via \`tensorboard --logdir logs/\`).
  ```

File: src/utils/callbacks.py [hunk 1]
```diff
@@ -15,6 +17,20 @@
...
+
+    Dispatches per-logger since Lightning loggers don't share a uniform image API:
+    WandbLogger exposes ``log_image``; TensorBoardLogger wraps a SummaryWriter
+    with ``add_figure``. Loggers without image support (CSV) are skipped.
+    """
+    for logger in trainer.loggers:
+        if isinstance(logger, WandbLogger):
+            logger.log_image(key=key, images=[fig], step=trainer.global_step)
+        elif isinstance(logger, TensorBoardLogger):
+            logger.experiment.add_figure(key, fig, global_step=trainer.global_step)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/utils/callbacks.py:43]: Good catch — fixed in 1df77b6. `_log_figure` now early-returns unless `trainer.is_global_zero`, with an added test `test_log_figure_is_noop_on_non_zero_rank` pinning the DDP-safe behavior. Docstring updated to explain why.

File: docs/reference/wandb-integration.md [hunk 2]
```diff
@@ -104,7 +114,7 @@ Logged via `self.log()` in each LightningModule:
 | `LearningRateMonitor` | Logs LR to Lightning logger                            | `configs/callbacks/lr_monitor.yaml`        |
 | `RichProgressBar`     | Terminal display only                                  | `configs/callbacks/rich_progress_bar.yaml` |
 | `ModelSummary`        | Prints param summary to console                        | `configs/callbacks/model_summary.yaml`     |
-| `PredictionWriter`    | Saves predictions to `.pt` files locally               | `src/utils/callbacks.py:307-342`           |
+| `PredictionWriter`    | Saves predictions to `.pt` files locally               | `src/utils/callbacks.py:320-355`           |
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/reference/wandb-integration.md]: Fair point — rather than refresh the numbers (which drifted twice in this PR alone), switched the whole table to Python-symbol references (`src/utils/callbacks.py::Class.method`) in 05fb7cb. Stable across future callback refactors.

File: docs/reference/wandb-integration.md [hunk 3]
```diff
@@ -87,14 +90,21 @@ Logged via `self.log()` in each LightningModule:
...
-| `PlotPositionalEncodingSimilarity` | `pos_enc_similarity` (image)     | `on_validation_epoch_end`                       | `src/utils/callbacks.py:135`     |
-| `PlotLearntProjection`             | `assignment`, `value` (images)   | `on_validation_epoch_end` or every N steps      | `src/utils/callbacks.py:259`     |
-| `LogPerParamMSE`                   | `per_param_mse/{name}` per param | `on_validation_epoch_end` (via `self.log_dict`) | `src/utils/callbacks.py:376-378` |
+| `PlotLossPerTimestep`              | `plot` (image)                   | `on_validation_epoch_end`                       | `src/utils/callbacks.py:91-93`   |
+| `PlotPositionalEncodingSimilarity` | `pos_enc_similarity` (image)     | `on_validation_epoch_end`                       | `src/utils/callbacks.py:148-150` |
+| `PlotLearntProjection`             | `assignment`, `value` (images)   | `on_validation_epoch_end` or every N steps      | `src/utils/callbacks.py:270-273` |
+| `LogPerParamMSE`                   | `per_param_mse/{name}` per param | `on_validation_epoch_end` (via `self.log_dict`) | `src/utils/callbacks.py:389-391` |
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/reference/wandb-integration.md]: Fair point — rather than refresh the numbers (which drifted twice in this PR alone), switched the whole table to Python-symbol references (`src/utils/callbacks.py::Class.method`) in 05fb7cb. Stable across future callback refactors.
